### ISSUE TITLE
VLA-JEPA default world-model loss leaks future frames into context via shared V-JEPA2 encoder pass

### ISSUE THREAD

**Issue by ravediamond** [huggingface/lerobot#4153]
### Ticket Type

🐛 Bug Report (Something isn't working)

### Environment & System Info

```Shell
Not runtime-environment specific — this is a logic bug in the loss computation, reproducible on any install. Found while reading src/lerobot/policies/vla_jepa/modeling_vla_jepa.py against the VLA-JEPA paper and the authors' reference implementation.
```

### Description

The default (always-on) world-model loss for VLA-JEPA leaks future frames into the context it predicts from, because context and target are both sliced out of a single shared encoder pass instead of being encoded separately.

- `VLAJEPAModel._world_model_loss` (`src/lerobot/policies/vla_jepa/modeling_vla_jepa.py:197-242`) runs the frozen V-JEPA2 encoder once over the whole video clip, then splits that one output tensor by index into `input_states` (context, fed to the predictor) and `gt_states` (the target it's supposed to predict).
- `torch.no_grad()` around the call keeps the encoder frozen, which is correct, but it only blocks gradients — it doesn't stop information flowing forward through attention during that single pass. 
- V-JEPA2's attention blocks are non-causal (`is_causal = False`, checked directly in `transformers/models/vjepa2/modeling_vjepa2.py`), so every token in the output, including the ones that end up in `input_states`, already attended to every frame in the clip — including the ones that end up in `gt_states`. The slicing happens after the fact, so it can't undo that.

This contradicts the paper's own design (Sun et al., arXiv:2602.10098v2):
- Eq. 1 defines the target state as a function of the frame at a single timestep, and the abstract is explicit that "a target encoder produces latent representations from future frames, while the student pathway sees only the current observation—future information is used solely as supervision targets, never as input." 
- As implemented, the context isn't leakage-free.

It affects every VLA-JEPA training run by default — not a rare edge case, not something you have to opt into.

### Context & Reproduction

Relevant code, `src/lerobot/policies/vla_jepa/modeling_vla_jepa.py:217-232`:

​```python
with torch.no_grad():
    video_embeddings = self.video_encoder.get_vision_features(pixel_values_videos=video_pixels)
    video_embeddings = torch.cat(torch.chunk(video_embeddings, chunks=v, dim=0), dim=2)

tubelet_size = self.video_encoder.config.tubelet_size
t_enc_total = self.config.num_video_frames // tubelet_size
t_enc_ctx = t_enc_total - 1
tokens_per_frame = video_embeddings.shape[1] // t_enc_total
input_states = video_embeddings[:, : tokens_per_frame * t_enc_ctx, :]
gt_states = video_embeddings[:, tokens_per_frame:, :]
​```

Also checked this against the paper authors' own reference implementation (`ginwind/VLA-JEPA`, `starVLA/model/framework/VLA_JEPA.py:224-232`) — same bug is present there too, same single-pass-then-slice pattern. So this isn't something the LeRobot port introduced; it's a faithful port of a bug in the original code. 

Filed upstream as ginwind/VLA-JEPA#29.

### Relevant logs or stack trace

```Shell

```

### Checklist

- [x] I have searched existing tickets to ensure this isn't a duplicate.
- [x] I am using the latest version of the `main` branch.
- [x] I have verified this is not an environment-specific problem.

### Additional Info / Workarounds

There's already precedent for exactly this type of fix in this codebase: PR #4116 ("Add opt-in multi-horizon training objective for VLA-JEPA world model") ran into the same problem in a newer, opt-in code path and fixed it there by encoding one causal raw-frame prefix per temporal position instead of one shared pass. But it's disabled by default and explicitly preserves "the legacy state dict, inference path, and single full-clip encoder call when the feature is disabled" — so the base path this issue is about was left untouched.

Suggested fix: 
- mirror #4116's approach — encode context and target from separate forward passes instead of one shared call, or at minimum make sure `input_states` comes from a raw-frame window that can't see any frame contributing to `gt_states`. 
- Will cost more encoder compute (per #4116's own numbers, roughly 4 encoder calls instead of 1 for the default 8-frame / tubelet-size-2 setup).

**Comment by ravediamond** [comment]
I'll keep this updated as the upstream issue (ginwind/VLA-JEPA#29) progresses — will comment here once the paper authors respond or land a fix.

**Comment by vdmaas98** [comment]
I reproduced the leak with a minimal script: a tiny randomly-initialised V-JEPA2 (transformers VJEPA2Model, 2 layers, CPU) driven through the same entry point the policy uses (get_vision_features), with the exact slicing logic of _world_model_loss:

- Legacy path (one shared pass, sliced into input_states and gt_states afterwards): perturbing only the last temporal position's raw frames changes input_states, with a max abs delta of 6.7e-3 on a tiny random encoder. The context contains future-frame information.

- Prefix-encoded path (the _causal_video_embeddings pattern from #4116, applied to the same positions): the identical perturbation produces a delta of exactly 0.0, while perturbing past frames still changes the context (a sanity check that the comparison isn't degenerate).

<summary>
Reproduction script (CPU, no checkpoints)</summary>

```python 
"""Demonstrate lerobot issue #4153: the default VLA-JEPA world-model loss leaks
future frames into the predictor context, and the #4116-style prefix encoding fixes it.

Reproduces the exact tensor logic of
`src/lerobot/policies/vla_jepa/modeling_vla_jepa.py::_world_model_loss`
(single shared V-JEPA2 pass, then index slicing) with a tiny randomly
initialised V-JEPA2 encoder on CPU, via the same public entry point the
policy uses (`get_vision_features`).

Run: python test_vla_jepa_leak.py
"""
import torch
from transformers.models.vjepa2.configuration_vjepa2 import VJEPA2Config
from transformers.models.vjepa2.modeling_vjepa2 import VJEPA2Model

torch.manual_seed(0)

NUM_FRAMES = 8  # config.num_video_frames default in VLAJEPAConfig

config = VJEPA2Config(
    crop_size=32,
    patch_size=16,
    frames_per_clip=NUM_FRAMES,
    tubelet_size=2,
    hidden_size=32,
    num_hidden_layers=2,
    num_attention_heads=2,
    mlp_ratio=2.0,
    pred_hidden_size=32,
    pred_num_attention_heads=2,
    pred_num_hidden_layers=1,
    pred_num_heads=2,
    pred_mlp_ratio=2.0,
)
encoder = VJEPA2Model(config).eval()

tubelet_size = config.tubelet_size
t_enc_total = NUM_FRAMES // tubelet_size          # 4 temporal positions
t_enc_ctx = t_enc_total - 1                       # 3 context positions

clip = torch.rand(1, NUM_FRAMES, 3, 32, 32)

# A perturbation strictly in the FUTURE relative to every context position:
# only the last temporal position's raw frames (frames 6..7) change.
clip_perturbed = clip.clone()
clip_perturbed[:, -tubelet_size:] = torch.rand_like(clip_perturbed[:, -tubelet_size:])


def legacy_context(video_pixels: torch.Tensor) -> torch.Tensor:
    """Exactly _world_model_loss's current context extraction (V=1 so the
    view-merge cat/chunk is a no-op): one shared pass, slice afterwards."""
    with torch.no_grad():
        video_embeddings = encoder.get_vision_features(pixel_values_videos=video_pixels)
    tokens_per_frame = video_embeddings.shape[1] // t_enc_total
    return video_embeddings[:, : tokens_per_frame * t_enc_ctx, :]  # input_states


def causal_context(video_pixels: torch.Tensor) -> torch.Tensor:
    """PR #4116's `_causal_video_embeddings` pattern: encode each temporal
    position from only its raw-frame prefix, keep that position's tokens."""
    positions = []
    with torch.no_grad():
        for position in range(t_enc_ctx):
            prefix_frames = (position + 1) * tubelet_size
            prefix = encoder.get_vision_features(
                pixel_values_videos=video_pixels[:, :prefix_frames]
            )
            tokens_per_position = prefix.shape[1] // (position + 1)
            positions.append(prefix[:, -tokens_per_position:])
    return torch.cat(positions, dim=1)


def max_delta(fn):
    return (fn(clip) - fn(clip_perturbed)).abs().max().item()


# --- 1. the bug: future-only perturbation changes the context -----------------
legacy_delta = max_delta(legacy_context)
print(f"legacy (shared pass + slice):  max |context delta| = {legacy_delta:.6f}")
assert legacy_delta > 1e-4, "expected the leak to be visible"

# --- 2. the fix: prefix encoding makes the context invariant to the future ----
causal_delta = max_delta(causal_context)
print(f"fixed  (per-position prefix):  max |context delta| = {causal_delta:.6f}")
assert causal_delta < 1e-6, "prefix-encoded context must not see future frames"

# --- 3. sanity: the fixed context still reacts to PAST frames -----------------
clip_past = clip.clone()
clip_past[:, :tubelet_size] = torch.rand_like(clip_past[:, :tubelet_size])
with torch.no_grad():
    past_delta = (causal_context(clip) - causal_context(clip_past)).abs().max().item()
print(f"sanity (perturb past frames):  max |context delta| = {past_delta:.6f}")
assert past_delta > 1e-4, "encoder must not be degenerate"

print("\nPASS: leak reproduced on the legacy path, absent on the prefix-encoded path.")
```

I'd like to implement the fix suggested in the issue (mirroring #4116's approach), since
nobody has picked it up yet. 

Concretely:
Reuse #4116's _causal_video_embeddings in the default _world_model_loss path: encode one raw-frame prefix per encoded temporal position, take each position's token block from its own prefix pass. Context becomes positions 0..T-2, targets positions 1..T-1, both from the same causal embeddings, consistent with what the multi-horizon path already does. The shift-by-one loss shape and logging stay identical. No state-dict changes (frozen encoder, same modules), and the view-merge layout is preserved.

Cost is the known one from #4116: T_enc encoder calls instead of 1 (4x at the default 8 frames / tubelet size 2). Frozen encoder, no_grad, so it is forward-only compute. The perturbation check above becomes the regression test: it asserts future-frame invariance of the context (fails on current main, passes with the fix) and past-frame sensitivity.

I saw the comment about waiting for the paper authors' response on ginwind/VLA-JEPA#29. Therefore I suggest putting the fix behind a config flag keeping the default faithful to the reference implementation.

**Comment by ravediamond** [comment]
Hello @vdmaas98,

Thanks for the detailed repro, nice write-up!

To clarify: #4116 doesn't fix this directly — it introduced the causal/prefix-encoding pattern that would, but hasn't been applied to the default path yet.

I'll take care of opening the fix PR for this, so no need to reimplement it separately. I'll link it here once it's up — would appreciate a review on the regression test if you have bandwidth.

On the config-flag idea: makes sense given the 4x encoder-call cost, I'll weigh it against just changing the default once I'm in the PR and lay out the tradeoff there.

I was basically waiting for either the original author in case they updated their code base (there are a few other open problems there), or a comment from the LeRobot core team. Given your repro, I think it's worth just moving forward now.

**Comment by vdmaas98** [comment]
Hi @ravediamond,

I put two suggestions in your draft PR. Have a look whenever you have the time!

**Comment by ravediamond** [comment]
Hello @vdmaas98,

Looks like the review's still pending on GitHub's side — mind hitting submit?

**Comment by vdmaas98** [comment]
Hi @ravediamond, 

Indeed forgot to hit submit. Just did it! Let me know what you think of the suggestions
